File: packages/agent/src/tool-dispatcher.ts
import type { ITool, ToolCall, ToolResult, ToolDescription, JsonObject } from '@cli-agent/core';
import {
  Registry,
  PermissionDeniedError,
  createChildLogger,
} from '@cli-agent/core';
import type { RunContext, AgentLogger } from '@cli-agent/core';
import { PermissionManager } from './permission.js';

/** Maximum characters in a single tool result output */
const MAX_OUTPUT_CHARS = 80_000;
const TRUNCATION_NOTICE = '\n\n... [output truncated — exceeded 80,000 characters]';

export class ToolDispatcher {
  private readonly toolRegistry: Registry<ITool>;
  private readonly permissionManager: PermissionManager;
  private readonly logger: AgentLogger;

  constructor(toolRegistry: Registry<ITool>, permissionManager: PermissionManager) {
    this.toolRegistry = toolRegistry;
    this.permissionManager = permissionManager;
    this.logger = createChildLogger('tool-dispatcher');
  }

  async dispatch(
    toolCall: ToolCall,
    context: RunContext
  ): Promise<ToolResult> {
    const tool = this.toolRegistry.tryGet(toolCall.name);
    if (!tool) {
      this.logger.warn({ toolName: toolCall.name }, 'Unknown tool');
      return {
        success: false,
        output: '',
        error: `Unknown tool: ${toolCall.name}`,
      };
    }

    let params: JsonObject;
    try {
      params = JSON.parse(toolCall.arguments) as JsonObject;
    } catch {
      this.logger.warn(
        { toolName: toolCall.name, arguments: toolCall.arguments.slice(0, 200) },
        'Invalid JSON in tool arguments',
      );
      return {
        success: false,
        output: '',
        error: `Invalid tool arguments for "${toolCall.name}": ${toolCall.arguments.slice(0, 100)}`,
      };
    }

    const permitted = await this.permissionManager.checkPermission(tool, params);
    if (!permitted) {
      this.logger.info({ toolName: toolCall.name }, 'Permission denied');
      throw new PermissionDeniedError(toolCall.name);
    }

    context.eventBus.emit('tool:start', { runId: context.runId, toolCall });

    let result: ToolResult;
    try {
      result = await tool.execute(params, context);
    } catch (error) {
      if (error instanceof PermissionDeniedError) throw error;
      const message = error instanceof Error ? error.message : String(error);
      result = { success: false, output: '', error: message };
    }

    const truncated = this.truncateResult(result);
    context.eventBus.emit('tool:end', { runId: context.runId, toolCall, result: truncated });
    return truncated;
  }

  async dispatchAll(
    toolCalls: readonly ToolCall[],
    context: RunContext
  ): Promise<ReadonlyMap<string, ToolResult>> {
    const abortedResult: ToolResult = {
      success: false,
      output: '',
      error: 'Operation aborted',
    };

    const entries = await Promise.all(
      toolCalls.map(async (toolCall): Promise<[string, ToolResult]> => {
        if (context.isAborted) {
          return [toolCall.id, abortedResult];
        }
        const result = await this.dispatch(toolCall, context);
        return [toolCall.id, result];
      })
    );

    return new Map(entries);
  }

  private truncateResult(result: ToolResult): ToolResult {
    if (result.output.length <= MAX_OUTPUT_CHARS) return result;
    this.logger.info(
      { originalLength: result.output.length, limit: MAX_OUTPUT_CHARS },
      'Tool output truncated',
    );
    return {
      ...result,
      output: result.output.slice(0, MAX_OUTPUT_CHARS) + TRUNCATION_NOTICE,
    };
  }

  getToolDescriptions(): ToolDescription[] {
    const descriptions: ToolDescription[] = [];
    for (const [, tool] of this.toolRegistry.getAll()) {
      descriptions.push(tool.describe());
    }
    return descriptions;
  }
}

File: sdk/python/chameleon/tool_dispatcher.py
"""
ToolDispatcher — Resolves tool calls from the LLM and executes them.
Python port of @cli-agent/agent ToolDispatcher.
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING

from .types import ToolCall, ToolDescription, ToolResult, RunContext

if TYPE_CHECKING:
    from .registry import Registry
    from .types import ITool

logger = logging.getLogger(__name__)

MAX_OUTPUT_CHARS = 80_000
TRUNCATION_NOTICE = "\n... [truncated — output exceeded 80 000 chars]"


class ToolDispatcher:
    def __init__(self, tool_registry: Registry[ITool]) -> None:
        self._registry = tool_registry

    async def dispatch(
        self,
        tool_call: ToolCall,
        context: RunContext,
    ) -> ToolResult:
        tool = self._registry.try_get(tool_call.name)
        if tool is None:
            return ToolResult(
                success=False,
                output="",
                error=f"Unknown tool: {tool_call.name}",
            )

        try:
            params = json.loads(tool_call.arguments) if tool_call.arguments else {}
        except json.JSONDecodeError as e:
            return ToolResult(
                success=False,
                output="",
                error=f"Invalid tool arguments JSON: {e}",
            )

        try:
            result = await tool.execute(params, context)
        except Exception as e:
            logger.exception("Tool %s execution error", tool_call.name)
            return ToolResult(
                success=False,
                output="",
                error=f"Tool execution error: {e}",
            )

        # Truncate oversized output
        if len(result.output) > MAX_OUTPUT_CHARS:
            result = ToolResult(
                success=result.success,
                output=result.output[:MAX_OUTPUT_CHARS] + TRUNCATION_NOTICE,
                error=result.error,
            )

        return result

    async def dispatch_parallel(
        self,
        tool_calls: list[ToolCall],
        context: RunContext,
    ) -> list[ToolResult]:
        """Execute multiple tool calls concurrently."""
        import asyncio

        tasks = [self.dispatch(tc, context) for tc in tool_calls]
        return list(await asyncio.gather(*tasks))

    def get_tool_descriptions(self) -> list[ToolDescription]:
        return [tool.describe() for tool in self._registry.get_all().values()]
